common/src/jni/main/cpp/conscrypt/native_crypto.cc

```diff
@@ -3801,6 +3799,466 @@ static jint NativeCrypto_EVP_AEAD_CTX_open_buf(JNIEnv* env, jclass, jlong evpAea
+    if (infoArray != nullptr) {
+        infoArrayElement = env->GetByteArrayElements(infoArray, 0);
+        info = reinterpret_cast<const uint8_t*>(infoArrayElement);
+        infoLen = env->GetArrayLength(infoArray);
+    }
+
+    bool isSetup = EVP_HPKE_CTX_setup_recipient(/* ctx= */ ctx,
+                                                /* key= */ key,
+                                                /* kdf= */ kdf,
+                                                /* aead= */ aead,
+                                                /* enc= */ reinterpret_cast<const uint8_t*>(enc),
+                                                /* enc_len= */ encLen,
+                                                /* info= */ info,
+                                                /* info_len= */ infoLen);
+
+    env->ReleaseByteArrayElements(encArray, enc, JNI_ABORT);
+    if (infoArrayElement != nullptr) {
+        env->ReleaseByteArrayElements(infoArray, infoArrayElement, JNI_ABORT);
+    }
+
+    if (!isSetup) {
+        conscrypt::jniutil::throwExceptionFromBoringSSLError(env, "EVP_HPKE_CTX_setup_recipient");
+        return nullptr;
+    }
+
+    return env->NewObject(conscrypt::jniutil::nativeRefHpkeCtxClass,
+                          conscrypt::jniutil::nativeRefHpkeCtxClass_constructor,
+                          reinterpret_cast<jlong>(ctx));
+}
+
+static jobjectArray NativeCrypto_EVP_HPKE_CTX_setup_sender(JNIEnv* env, jclass, jint kemValue,
+                                                           jint kdfValue, jint aeadValue,
+                                                           jbyteArray publicKeyArray,
+                                                           jbyteArray infoArray) {
+    CHECK_ERROR_QUEUE_ON_RETURN;
+    JNI_TRACE("EVP_HPKE_CTX_setup_sender(%d, %d, %d, %p, %p)", kemValue, kdfValue, aeadValue,
+              publicKeyArray, infoArray);
+
+    EVP_HPKE_CTX* ctx = EVP_HPKE_CTX_new();
+    const EVP_HPKE_KEM* kem = getHpkeKem(env, kemValue);
+    if (kem == nullptr) {
+        return nullptr;
+    }
+    const EVP_HPKE_KDF* kdf = getHpkeKdf(env, kdfValue);
+    if (kdf == nullptr) {
+        return nullptr;
+    }
+    const EVP_HPKE_AEAD* aead = getHpkeAead(env, aeadValue);
+    if (aead == nullptr) {
+        return nullptr;
+    }
+    if (publicKeyArray == nullptr) {
+        conscrypt::jniutil::throwNullPointerException(env, "publicKeyArray == null");
+        return {};
+    }
+
+    jbyte* peer_public_key = env->GetByteArrayElements(publicKeyArray, 0);
+
+    jbyte* infoArrayElement = nullptr;
+    const uint8_t* info = nullptr;
+    size_t infoLen = 0;
+    if (infoArray != nullptr) {
+        infoArrayElement = env->GetByteArrayElements(infoArray, 0);
+        info = reinterpret_cast<const uint8_t*>(infoArrayElement);
+        infoLen = env->GetArrayLength(infoArray);
+    }
+
+    size_t encapsulatedSharedSecretLen;
+    std::vector<uint8_t> encapsulatedSharedSecret(EVP_HPKE_MAX_ENC_LENGTH);
```

Comment:
No need to put this in a `std::vector` and waste a heap allocation. The expectation is that you stack-allocate:

```
uint8_t enc[EVP_HPKE_MAX_ENC_LENGTH];
```

Reply:
Done. Code updated to use `uint8_t enc[EVP_HPKE_MAX_ENC_LENGTH];` instead.